crates/uv-resolver/src/pubgrub/dependencies.rs

```diff
@@ -162,22 +169,28 @@ impl PubGrubRequirement {
                 parsed_url,
                 verbatim: verbatim_url.clone(),
             }),
+            local: None,
         })
     }
 
     fn from_registry_requirement(
         specifier: &VersionSpecifiers,
         extra: Option<ExtraName>,
         requirement: &Requirement,
-        locals: &Locals,
+        fork_locals: &ForkLocals,
     ) -> Result<PubGrubRequirement, ResolveError> {
-        // If the specifier is an exact version, and the user requested a local version that's
-        // more precise than the specifier, use the local version instead.
-        let version = if let Some(expected) = locals.get(&requirement.name) {
+        // If the specifier is an exact version and the user requested a local version for this
+        // fork that's more precise than the specifier, use the local version instead.
+        let local = fork_locals
+            .get(&requirement.name)
+            .cloned()
+            .or_else(|| locals::from_source(&requirement.source));
```

Comment:
This is a little unfortunate because now we call `from_source` for every dependency (including transitive dependencies), but it's harder to do this in the resolver because we don't want to set `local` for extras or markers.